### PR TITLE
Change focused widget with arrow keys

### LINKED ISSUE #1250: Gampad and arrow-keys navigation support
**Is your feature request related to a problem? Please describe.**

I'd like to be able to navigate on the gui elments using a gamepad

### PR DESCRIPTION
**Pull request by TimonPost**
We have the use case of gamepad control for a user-facing game menu. This PR makes it possible to change widget focus with the keyboard arrows. It allows integrations to convert joysticks to arrow keys in order to toggle between menu items. It does not add support for gamepad events in egui as this seems like a bigger topic that I desire not to touch. 

Initially, I examine the currently focused widget to determine the one with the highest dot product when compared to the desired focus direction. Subsequently, I identify the widget that is nearest to the current one. This process enables the selection of the widget in the desired direction which not only possesses the smallest angle relative to the current widget but also maintains the closest proximity.

The widget selection code can be improved over time ofcourse, there are some potential issues right now:
- This solution has no sense of the window it exists in, so if a focusable widget in another window is a better candidate it will pick that one.  
- If I click in an input box the `Focus::interested_in_focus` is not called and thus I do not know what the interactable widgets and their rectangles are. I only know this after using a key like arrow or tab (+shift). 

Closes #1250

### CODE REVIEW

File: crates/egui/src/memory.rs [hunk 1]
```diff
@@ -322,31 +355,121 @@ impl Focus {
...
+            Vec::with_capacity(self.focus_widgets_cache.len());
+
+        for (widget_id, widget_rect) in &mut self.focus_widgets_cache {
+            if Some(*widget_id) == self.id {
+                continue;
+            }
+
+            let candidate_widget_pos = widget_rect.left_center();
+
+            let current_to_candidate = candidate_widget_pos - current_widget_pos;
```
Comment (emilk) [inline crates/egui/src/memory.rs:423]: This could be improved. Instead of taking an arbitrary point (`left_center`) in the rect, we should consider the full rects.

For instance, these right-aligned rects should be considered to be perfectly above each other:

```
[ first widget ]
     [ another ]
        [ last ]
```

We can accomplish this with a function that checks the x-ranges and y-ranges and if they overlap, returns 0 for that axis.

```rs
/// * negative if `a` is strictly left of `b`
/// * positive if `a` is strictly right of `b`
/// * zero if the ranges overlap
fn range_diff(a: Rangef, b: Rangef) -> f32 {
    if a.max < b.min {
        a.max - b.min
    } else if b.max < a.min {
        a.min - b.max
    } else {
        0.0 // overlaps
    }
}

…

let current_to_candate = vec2(
    range_diff(selected_rect.x_range(), candidate_rect.x_range(),
    range_diff(selected_rect.y_range(), candidate_rect.y_range(),
);
```

I'm not a 100% this is an improvement, but I believe so.

File: examples/hello_world/src/main.rs [hunk 2]
```diff
@@ -17,14 +17,18 @@ fn main() -> Result<(), eframe::Error> {
 
 struct MyApp {
     name: String,
+    name2: String,
     age: u32,
+    age2: u32,
```
Comment (emilk) [inline examples/hello_world/src/main.rs:22]: please revert the changes in this file. Use `egui_demo_app` (`cargo r`) for a better test of your code!

**Review by emilk — CHANGES_REQUESTED** [review]
Exciting!

The navigation works okish, but I left some ideas on how to improve it

File: crates/egui/src/memory.rs [hunk 3]
```diff
@@ -322,31 +355,121 @@ impl Focus {
...
+            if let Some(new_rect) = new_rects.get(id) {
+                *old_rect = *new_rect;
+                true // Keep the item
+            } else {
+                false // Remove the item
+            }
+        });
+
+        let current_focus_widget_rect = self.focus_widgets_cache.get(&focus_id).unwrap();
+        let current_widget_pos = current_focus_widget_rect.left_center();
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply crates/egui/src/memory.rs:411]: I am not entirely sure what the best metrics is. I think going from center, left, right, top down all have their consequences. 

Left The center might change based on the contents, the left center will not change based on content. I had the issue were i wanted to go to the widget to the right but the widget to the right top was closer center wise so I went up right instead.

File: crates/egui/src/memory.rs [hunk 4]
```diff
@@ -322,31 +355,121 @@ impl Focus {
...
+            // Check if the widget has the right dot product and length.
+            let focus_direction = match self.focus_direction {
+                FocusDirection::Up => Vec2::UP,
+                FocusDirection::Right => Vec2::RIGHT,
+                FocusDirection::Down => Vec2::DOWN,
+                FocusDirection::Left => Vec2::LEFT,
+                _ => {
+                    continue;
+                }
+            };
```
[the remark this answers is not in the document]
Comment (TimonPost) [reply crates/egui/src/memory.rs:447]: If `current_to_candidate` is perpendicular to `focus_direction` it is 0. If `current_to_candidate` is parallel to `focus_direction` it is 0. `aligns_with_focus_direction` exists to prevent picking the perpendicular case. I do not see how this change solves this problem. Can you elaborate?

[the remark this answers is not in the document]
Comment (TimonPost) [reply crates/egui/src/memory.rs]: Yes your right!

**Comment by TimonPost** [comment]
I updated the selection metric but it is still not perfect.
```
focus_candidates.sort_by(|(_, dot1, len1), (_, dot2, len2)| {
    (len1 / dot1).partial_cmp(&(len2 / dot2)).unwrap()
});
focus_candidates.sort_by(|(_, dot1, _), (_, dot2, _)| dot2.partial_cmp(dot1).unwrap());
```

Not sorting by dot product afterwards results in close by widgets to be picked evenwhile they have a big angle. I am not sure how perfect this feature has to be. 

An other issue is that sometimes it's hard to tell where focus went to as it might be a window behind the current window, or some widget in a different panel. 

The rectangle thing does seem to improve the overall behavior.

File: crates/emath/src/range.rs [hunk 5]
```diff
@@ -49,12 +49,48 @@ impl Rangef {
...
+        self.min + (self.span() / 2.0)
+    }
+
+    /// Returns the similarity between `self` and `other`.
+    ///
+    ///  * negative if `a` is left of `b`
+    /// * positive if `a` is right of `b`
+    /// * zero if the ranges overlap significantly
+    #[inline]
+    pub fn similarity(self, other: Rangef) -> f32 {
```
Comment (emilk) [inline crates/emath/src/range.rs:64]: This is a heuristic specific to the keyboard problem and doesn't belong in this file

File: crates/emath/src/range.rs [hunk 6]
```diff
@@ -49,12 +49,48 @@ impl Rangef {
...
+
     #[inline]
     #[must_use]
     pub fn contains(self, x: f32) -> bool {
         self.min <= x && x <= self.max
     }
 
+    /// Returns if `self` contains `other` for at least 50%.
+    #[inline]
+    pub fn contains_range(&self, other: Rangef) -> bool {
```
Comment (emilk) [inline crates/emath/src/range.rs:80]: This is a very surprising definition of `contains_range`. Either change it to `self.contains(other.min) && self.contains(other.max)`, or move it to where it is being used.

Comment (TimonPost) [reply crates/emath/src/range.rs:80]: Was trying to get the overlap of more than 50 percent to work. So that self looks if it overlaps at least 50% with other. Changing it to self.contains(other.min) && self.contains(other.max) would restrict it to a full overlap where other is fully within self?

Comment (emilk) [reply crates/emath/src/range.rs:80]: the 50% overlap may be a good idea for what you are trying to do, but it is not a good idea for a `pub` function called `contains_range`. Call it `has_significant_overlap` or something and move it to where it is used (`crates/egui/src/memory.rs`)

Comment (emilk) [reply crates/emath/src/range.rs:80]: the 50% overlap may be a good idea for what you are trying to do, but it is not a good idea for a `pub` function called `contains_range`. Call it `has_significant_overlap` or something and move it to where it is used (`crates/egui/src/memory.rs`)

**Review by emilk — APPROVED** [review]

**Comment by emilk** [comment]
There is still plenty of room for improvement, but this is a good start!

**Comment by Jakkestt** [comment]
>It allows integrations to convert joysticks to arrow keys in order to toggle between menu items.
How do you do that?

**Comment by TimonPost** [comment]
Gotta make some custom logic to achieve that. E.g detect the cardinal directions of the joystick and map them to key combo's
